frontend/javascripts/test/sagas/volumetracing/volumetracing_saga.spec.ts

```diff
@@ -394,3 +420,72 @@ test("VolumeTracingSaga should ignore brush action when busy (saga test)", (t) =
     take("START_EDITING"),
   );
 });
+
+test("VolumeTracingSaga should pin an active mapping upon first volume annotation", (t) => {
+  const saga = editVolumeLayerAsync();
+  saga.next();
+  saga.next();
+  expectValueDeepEqual(t, saga.next(true), take("START_EDITING"));
+  saga.next(startEditingAction);
+  saga.next({
+    isBusy: false,
+  });
+  saga.next(volumeTracing);
+  saga.next(OverwriteModeEnum.OVERWRITE_ALL);
+  saga.next(AnnotationToolEnum.BRUSH);
+  saga.next(false);
+  // pass labeled resolution
+  saga.next({
+    resolution: [1, 1, 1],
+    zoomStep: 0,
+  });
+  // Test whether nested saga ensureMaybeActiveMappingIsPinned is called.
+  expectValueDeepEqual(
+    t,
+    saga.next(ACTIVE_CELL_ID),
+    call(ensureMaybeActiveMappingIsPinned, volumeTracing),
+  );
+});
+
+test("ensureMaybeActiveMappingIsPinned should pin an existing mapping to the annotation", (t) => {
+  const activeMappingByLayer = { [volumeTracing.tracingId]: dummyActiveMapping };
+  const saga = ensureMaybeActiveMappingIsPinned(volumeTracing);
+  saga.next();
+  expectValueDeepEqual(
+    t,
+    saga.next({ [volumeTracing.tracingId]: dummyActiveMapping }),
+    call(askUserForPinningActiveMapping, volumeTracing, activeMappingByLayer),
+  );
+  t.true(saga.next().done);
+});
+
+test("ensureMaybeActiveMappingIsPinned should pin 'no mapping' in case no mapping is active.", (t) => {
+  const saga = ensureMaybeActiveMappingIsPinned(volumeTracing);
+  saga.next();
+  expectValueDeepEqual(t, saga.next({}), put(VolumeTracingActions.setMappingIsPinnedAction()));
+  t.true(saga.next().done);
+});
+
+test("ensureMaybeActiveMappingIsPinned should pin 'no mapping' in case a mapping is active but disabled.", (t) => {
+  const jsonDummyMapping = { ...dummyActiveMapping, mappingStatus: MappingStatusEnum.DISABLED };
+  const saga = ensureMaybeActiveMappingIsPinned(volumeTracing);
+  saga.next();
+  expectValueDeepEqual(
+    t,
+    saga.next({ [volumeTracing.tracingId]: jsonDummyMapping }),
+    put(VolumeTracingActions.setMappingIsPinnedAction()),
+  );
+  t.true(saga.next().done);
+});
+
+test("ensureMaybeActiveMappingIsPinned should pin 'no mapping' in case a JSON mapping is active.", (t) => {
+  const jsonDummyMapping = { ...dummyActiveMapping, mappingType: "JSON" };
+  const saga = ensureMaybeActiveMappingIsPinned(volumeTracing);
+  saga.next();
+  expectValueDeepEqual(
+    t,
+    saga.next({ [volumeTracing.tracingId]: jsonDummyMapping }),
+    put(VolumeTracingActions.setMappingIsPinnedAction()),
+  );
+  t.true(saga.next().done);
+});
```

Comment:
A still open question to me is how the pinning should behave in case the user has an active JSON mapping. As JSON mappings cannot be manipulated (changes to it are not persisted), should the state that it is active be pinned, or should "not active mapping" be pinned? For consistency reasons, the state of whether a mapping and which mapping is pinned should be determined by the first volume annotation action.

Here are the conversational bits of this pr discussing this:
> @fm3 How do you think we should handle JSON Mappings? Because the code says that changes to these mappings are not persisted? Should they also be pinned (this is how it is currently implemented in the frontend)

> Yes, probably. Did you test if that works as expected? I haven’t worked with JSON mapings in a while

> When activating a JSON mapping, and then brushing, it is pinned regardless (although the PR description states the pinning is only supported for HDF5 mappings). When reloading the annotation the mapping is disabled, but pinned, and cannot be activated or changed - which is unexpected. I'd avoid pinning JSON mappings in the first place if it's not properly supported and leads to weird behavior.

My thoughts on this would be to ensure that a mapping state is pinned, that the state of "no mapping is active" should be pinned to the annotation once the user annotates in some way while a JSON Mapping is active.

=> One open TODO for this (in my opinion) would be, to also disable / deactivate the JSON mapping once any volume annotation is done. Because if it stays active and the user annotates a little and then reloads, the user seems different results as the JSON mapping was previously still active but it cannot be enabled anymore after the reload.

@daniel-wer @philippotto What are your opinions on this?

Reply:
I see two main options to make the user experience consistent:

1) Pin the JSON mapping the same way that agglomerate mappings are pinned. Reactivate the mapping after a reload, make it unchangeable, etc.

2) Forbid volume annotation actions while a JSON mapping is active. (Disabling the mapping once any volume annotation is done is potentially confusing imo)

Since JSON mappings are a very niche use case, I'd go with the option that's the easiest to implement.

Reply:
Just as a side note: there are also dynamic mappings (e.g., by merger mode). I'm not sure whether these are interpreted as JSON mappings in some cases. Locking these forever would be unusual in the context of the merger mode, I think.
I'd probably prefer (2).

Reply:
:+1: I implemented (2) and it was pretty easy to integrate this.